pkg/ingest/aggregator.go

```diff
@@ -0,0 +1,320 @@
+	}
+}
+
+func newAggrDf(ao AggrsOptions) AggrDf {
+	return AggrDf{seriesRecordSets: make(map[uint64]*SeriesRecordSet)}
+}
+
+// Ingest a single chunk provided via an iterator. For a specific searies, We
+// assume the iterator returns values ordered by the timestamp
+// The iterator is expected to already be at the point of the first sample after as.sampleStart
+func (a *aggregator) ingestChunk(as *aggregatedSeries, i input.ChunkIterator) (*aggregatedSeries, error) {
+	var (
+		ts int64
+		v  float64
+		t  time.Time
+	)
+	for {
+		ts, v = i.At()
+		t = timestamp.Time(ts)
+		if t.Before(as.sampleStart) {
+			return as, errors.Errorf("Chunk timestamp %s is less than the sampleStart %s", t, as.sampleStart)
+		}
+		if t.After(as.sampleEnd) {
+			as = a.finalizeSample(as, t)
+		}
+
+		if as.count == 0 {
+			as.minTime = t
+			as.maxTime = t
+			as.min = v
+			as.max = v
+		}
+		if as.maxTime.After(t) {
+			return as, errors.Errorf("Incoming chunks are not sorted by timestamp: expected %s after %s", t, as.maxTime)
+		}
+		as.maxTime = t
+		as.count += 1
+		as.sum += v
+		if as.max < v {
+			as.max = v
+		}
+		if as.min > v {
+			as.min = v
+		}
+		if !i.Next() {
+			break
+		}
+	}
+	return as, nil
+}
+
+func (a *aggregator) addSeriesToDf(as *aggregatedSeries) {
+	rs, ok := a.df.seriesRecordSets[as.hash]
+	if !ok {
+		rs = a.df.addRecordSet(as.labels)
+	}
+
+	vals := map[string]interface{}{
+		"_sample_start": as.sampleStart,
+		"_sample_end":   as.sampleEnd,
+		"_min_time":     as.minTime,
+		"_max_time":     as.maxTime,
+	}
+
+	for _, l := range as.labels {
+		if l.Name == "__name__" {
+			continue
+		}
+		vals[l.Name] = l.Value
+	}
+
+	if a.aggrsOptions.Count.Enabled {
+		vals[a.aggrsOptions.Count.Column] = as.count
+	}
+	if a.aggrsOptions.Sum.Enabled {
+		vals[a.aggrsOptions.Sum.Column] = as.sum
+	}
+	if a.aggrsOptions.Min.Enabled {
+		vals[a.aggrsOptions.Min.Column] = as.min
+	}
+	if a.aggrsOptions.Max.Enabled {
+		vals[a.aggrsOptions.Max.Column] = as.max
+	}
+	rs.Records = append(rs.Records, Record{Values: vals})
+}
+
+// Add the active aggregated series into the final dataframe when we've reached the
+// sample end time. Returns pointer to a new instance of the aggregatedSeries
+func (a *aggregator) finalizeSample(as *aggregatedSeries, nextT time.Time) *aggregatedSeries {
+	if as.count > 0 {
+		a.addSeriesToDf(as)
+	}
+
+	// calculate the next sample cycle to contain the nextT time. First calculate how many
+	// whole resolution cycles are between current sampleStart and nextT and then add
+	// those cycles to the current sampleStart
+	nextSampleCycle := ((nextT.Unix() - as.sampleStart.Unix()) / (int64)(a.Resolution/time.Second))
+	nextSampleStart := as.sampleStart.Add((time.Duration(nextSampleCycle)) * a.Resolution)
+
+	as = &aggregatedSeries{
+		labels:      as.labels,
+		hash:        as.hash,
+		sampleStart: nextSampleStart,
+		sampleEnd:   nextSampleStart.Add(a.Resolution),
+	}
+	a.activeSeries[as.hash] = as
+	return as
```

Comment:
do we need to return it?

Reply:
Yes because we actually create a new series and in case the caller (such as in `insgestChunk`) stores the reference to the as, they should be able to just update to the new one.